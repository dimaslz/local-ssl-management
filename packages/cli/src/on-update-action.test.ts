import fs from "fs";
import shell from "shelljs";

import generateProxyImage from "./generate-proxy-image";
import onUpdateAction from "./on-update-action";
import validatePort from "./utils/validate-port";

vi.mock("./generate-proxy-image");
vi.mock("./utils/validate-port");
vi.mock("fs");
vi.mock("shelljs");
vi.mock("@dimaslz/local-ssl-management-core", () => ({
  mkcert: vi.fn(),
  getLocalIP: vi.fn(() => "192.168.0.0"),
}));
vi.mock("./list-container");
vi.mock("path", () => ({
  default: {
    resolve: () => "/root/path",
  },
}));

vi.mock("chalk", async () => ({
  default: {
    green: vi.fn((v) => v),
    red: vi.fn((v) => v),
  },
}));

describe("On update action", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("failures", () => {
    test("update domain by wrong domain key", () => {
      const domain = "foo-domain.com";
      const port = 3333;

      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3000,
            },
          ],
          null,
          2,
        ),
      );
      vi.spyOn(shell, "exit").mockImplementation(() => {
        throw new Error();
      });

      expect(() => {
        onUpdateAction(domain, { port });
      }).toThrow();

      expect(shell.echo).toBeCalledWith(
        '\n[Error] - Domain "foo-domain.com" does not exists\n',
      );
      expect(shell.exit).toBeCalledWith(1);
    });

    test("update domain by wrong id", () => {
      const domain = "48d1a85c-377a-40ef-8a82-d1405f7a0722";
      const port = 3333;

      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3000,
            },
          ],
          null,
          2,
        ),
      );
      vi.spyOn(shell, "exit").mockImplementation(() => {
        throw new Error();
      });

      expect(() => {
        onUpdateAction(domain, { port });
      }).toThrow();

      expect(shell.echo).toBeCalledWith(
        '\n[Error] - Domain with key "48d1a85c-377a-40ef-8a82-d1405f7a0722" does not exists\n',
      );
      expect(shell.exit).toBeCalledWith(1);
    });
  });

  describe("success", () => {
    test("update domain by domain key", () => {
      const domain = "some-domain.com";
      const port = 3333;

      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3000,
            },
          ],
          null,
          2,
        ),
      );

      onUpdateAction(domain, { port });

      expect(validatePort).toBeCalled();

      expect(fs.writeFileSync).toHaveBeenCalledWith(
        "/root/path/.local-ssl-management/config.json",
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3333,
            },
          ],
          null,
          2,
        ),
      );

      expect(shell.echo).toBeCalledTimes(2);
      expect(shell.echo).nthCalledWith(
        1,
        "\n[Success] - ðŸŽ‰ Domain removed succesful.\n",
      );
      expect(shell.echo).nthCalledWith(
        2,
        "\n[Action] - ðŸ”„ Updating proxy image.\n",
      );

      expect(generateProxyImage).toBeCalled();
    });

    test("update domain by domain id", () => {
      const domain = "48d1a85c-377a-40ef-8a82-d1405f7a074f";
      const port = 3333;

      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3000,
            },
          ],
          null,
          2,
        ),
      );

      onUpdateAction(domain, { port });

      expect(validatePort).toBeCalled();

      expect(fs.writeFileSync).toHaveBeenCalledWith(
        "/root/path/.local-ssl-management/config.json",
        JSON.stringify(
          [
            {
              id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
              domain: "some-domain.com",
              port: 3333,
            },
          ],
          null,
          2,
        ),
      );

      expect(shell.echo).toBeCalledTimes(2);
      expect(shell.echo).nthCalledWith(
        1,
        "\n[Success] - ðŸŽ‰ Domain removed succesful.\n",
      );
      expect(shell.echo).nthCalledWith(
        2,
        "\n[Action] - ðŸ”„ Updating proxy image.\n",
      );

      expect(generateProxyImage).toBeCalled();
    });
  });
});
