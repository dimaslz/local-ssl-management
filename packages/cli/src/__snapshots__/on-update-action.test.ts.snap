// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`On update action > success > update location > update domain by domain id 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/root/path/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "5ac20bc9-4d7c-43e5-a1ae-a8f09866a7e1",
        "port": "3000",
        "location": "/app-name"
      }
    ]
  }
]",
    ],
    [
      "/root/path/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/app-name(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/root/path/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On update action > success > update location > update domain by domain key 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/root/path/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "5ac20bc9-4d7c-43e5-a1ae-a8f09866a7e1",
        "port": "3000",
        "location": "/app-name"
      }
    ]
  }
]",
    ],
    [
      "/root/path/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/app-name(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/root/path/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On update action > success > update location and port > update domain by domain id 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/root/path/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "5ac20bc9-4d7c-43e5-a1ae-a8f09866a7e1",
        "port": "4000",
        "location": "/app-name"
      }
    ]
  }
]",
    ],
    [
      "/root/path/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/app-name(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:4000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/root/path/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On update action > success > update port > update domain by domain id 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/root/path/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "5ac20bc9-4d7c-43e5-a1ae-a8f09866a7e1",
        "port": "3333",
        "location": "/"
      }
    ]
  }
]",
    ],
    [
      "/root/path/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3333/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/root/path/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On update action > success > update port > update domain by domain key 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/root/path/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "5ac20bc9-4d7c-43e5-a1ae-a8f09866a7e1",
        "port": "3333",
        "location": "/"
      }
    ]
  }
]",
    ],
    [
      "/root/path/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3333/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/root/path/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
